### DISCUSSION TITLE
Regrid_Data_Plane Error with Opening File

### DISCUSSION THREAD

**Discussion by cjmelick** [dtcenter/METplus#1416] [File I/O] [answered]
Using Regrid_Data_Plane, I'm trying to regrid model data at a coarse resolution on to a higher resolution limited area domain used by the NOHRSC Snowfall Analysis and I get an error saying parse_vx_grid() -> can't open file "sfav2_CONUS_24h_2022020312.nc".   The netCDF file permissions are readable and I can display the contents using ncdump.    Do you have any idea why I would be getting this type of error?   

Thanks!
Chris

**Comment by JohnHalleyGotway** [comment]
@cjmelick I see you're having trouble reading a NetCDF data file into MET. You are getting an error stating:
```
parse_vx_grid() -> can't open file "sfav2_CONUS_24h_2022020312.nc"
```

To replicate I went hunting around and found a similar file available here:
https://www.nohrsc.noaa.gov/snowfall_v2/

I ran a regrid_data_plane command for MET version 10.0.0 but got a slightly different error message:
```
regrid_data_plane sfav2_CONUS_24h_2022020712.nc G001 regrid.nc -field 'name="Data"; level="(*,*)";'
ERROR  : 
ERROR  : process_data_file() -> "sfav2_CONUS_24h_2022020712.nc" not a valid data file
ERROR  :
```

Running "ncdump -h" I do see that this claims to be a CF-compliant NetCDF file.
```
string :Conventions = "CF-1.6" ;
```

Rerunning using the "file_type = NETCDF_NCCF;" to explicitly tell it to interpret it as such does not produce much better results... and this time I switched to MET's plot_data_plane, which is really useful when reading new datasets:
```
plot_data_plane sfav2_CONUS_24h_2022020712.nc plot.ps 'name="Data"; level="(*,*)"; file_type=NETCDF_NCCF;'
...
ERROR  : 
ERROR  : NcCfFile::get_grid_from_grid_mapping() -> Cannot extract grid mapping variable ((nul)) from netCDF file.
ERROR  : 
```

I strongly suspect the issue stems from this data file using the NetCDF string data type. If it contained character arrays instead, I suspect it'd work just fine in MET. I ran the following ncdump and ncgen commands to test. They basically strip out all references to "string":
```
ncdump sfav2_CONUS_24h_2022020712.nc | sed 's/string //g' > sfav2_CONUS_24h_2022020712.ncdump
ncgen -o sfav2_CONUS_24h_2022020712-char.nc sfav2_CONUS_24h_2022020712.ncdump
```

And then I plotted the result:
```
plot_data_plane sfav2_CONUS_24h_2022020712-char.nc plot.ps 'name="Data"; level="(*,*)";'
```

And it works. Here's the resulting image:
<img width="596" alt="Screen Shot 2022-02-07 at 12 25 05 PM" src="https://user-images.githubusercontent.com/21087144/152857607-525d5383-c261-45e6-bed2-aa37fe420a1a.png">

I think there's 3 points to clarify here.

1. Is this the actual problem you're experiencing? Or is there really some permissions issue preventing MET from reading the input file in the first place?

2. Note that the successful plot_data_plane command did print this warning message:
```
WARNING: NcCfFile::open() -> could not determine the valid time, using 0.
```

While I see that the "Data" variable does include some timing information, it is not encoded in a CF-compliant way. If you have any control over the creation of this data, I'd recommend following the conventions for the "[Time Coordinate](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.9/cf-conventions.html#time-coordinate)".

3. Does specifying character arrays as NetCDF strings fall within the CF-convention? Meaning, should the MET library code be enhanced to support this convention? 

@hsoh-u can you please investigate? Should MET be able to handle "string" variable attributes, like this?
```
	double lat(lat) ;
		string lat:long_name = "latitude" ;
		string lat:units = "degrees_north" ;
		string lat:standard_name = "latitude" ;
		lat:resolution = 0.04 ;
		lat:origin_offset = 0.02 ;
		string lat:bounds = "lat_bounds" ;
```

**Comment by cjmelick** [comment]
Hi John,

Thanks for taking the time to make some modifications to the netCDF file
and successfully displaying the data!  I sometimes use the NOHRSC Snowfall
Analysis to do verification of snowfall amounts.   I don't have any control
over the creation process -- I would think that NOAA would follow the
CF-compliant guidelines.   I suppose I could do some post-processing like
you did to get things to work right.

As for your specific questions,

1.).  Great question.   I know that there should not be any permission
issue with the netCDF file as I was able to run ncdump just fine.    I will
have to try tomorrow what you suggested before running regrid_data_plane.
2 and 3.). Not sure.   I wish I had more control.    Again, I could reach
out to the NOHRSC group and/or make some of my own corrections in
post-processing (assuming this is the problem)

More to follow soon,
Chris

On Mon, Feb 7, 2022 at 1:31 PM johnhg ***@***.***> wrote:

> @cjmelick <https://github.com/cjmelick> I see you're having trouble
> reading a NetCDF data file into MET. You are getting an error stating:
>
> parse_vx_grid() -> can't open file "sfav2_CONUS_24h_2022020312.nc"
>
> To replicate I went hunting around and found a similar file available here:
> https://www.nohrsc.noaa.gov/snowfall_v2/
>
> I ran a regrid_data_plane command for MET version 10.0.0 but got a
> slightly different error message:
>
> regrid_data_plane sfav2_CONUS_24h_2022020712.nc G001 regrid.nc -field 'name="Data"; level="(*,*)";'
> ERROR  :
> ERROR  : process_data_file() -> "sfav2_CONUS_24h_2022020712.nc" not a valid data file
> ERROR  :
>
> Running "ncdump -h" I do see that this claims to be a CF-compliant NetCDF
> file.
>
> string :Conventions = "CF-1.6" ;
>
> Rerunning using the "file_type = NETCDF_NCCF;" to explicitly tell it to
> interpret it as such does not produce much better results... and this time
> I switched to MET's plot_data_plane, which is really useful when reading
> new datasets:
>
> plot_data_plane sfav2_CONUS_24h_2022020712.nc plot.ps 'name="Data"; level="(*,*)"; file_type=NETCDF_NCCF;'
> ...
> ERROR  :
> ERROR  : NcCfFile::get_grid_from_grid_mapping() -> Cannot extract grid mapping variable ((nul)) from netCDF file.
> ERROR  :
>
> I strongly suspect the issue stems from this data file using the NetCDF
> string data type. If it contained character arrays instead, I suspect it'd
> work just fine in MET. I ran the following ncdump and ncgen commands to
> test. They basically strip out all references to "string":
>
> ncdump sfav2_CONUS_24h_2022020712.nc | sed 's/string //g' > sfav2_CONUS_24h_2022020712.ncdump
> ncgen -o sfav2_CONUS_24h_2022020712-char.nc sfav2_CONUS_24h_2022020712.ncdump
>
> And then I plotted the result:
>
> plot_data_plane sfav2_CONUS_24h_2022020712-char.nc plot.ps 'name="Data"; level="(*,*)";'
>
> And it works. Here's the resulting image:
> [image: Screen Shot 2022-02-07 at 12 25 05 PM]
> <https://user-images.githubusercontent.com/21087144/152857607-525d5383-c261-45e6-bed2-aa37fe420a1a.png>
>
> I think there's 3 points to clarify here.
>
>    1.
>
>    Is this the actual problem you're experiencing? Or is there really
>    some permissions issue preventing MET from reading the input file in the
>    first place?
>    2.
>
>    Note that the successful plot_data_plane command did print this
>    warning message:
>
> WARNING: NcCfFile::open() -> could not determine the valid time, using 0.
>
> While I see that the "Data" variable does include some timing information,
> it is not encoded in a CF-compliant way. If you have any control over the
> creation of this data, I'd recommend following the conventions for the "Time
> Coordinate
> <https://cfconventions.org/Data/cf-conventions/cf-conventions-1.9/cf-conventions.html#time-coordinate>
> ".
>
>    1. Does specifying character arrays as NetCDF strings fall within the
>    CF-convention? Meaning, should the MET library code be enhanced to support
>    this convention?
>
> @hsoh-u <https://github.com/hsoh-u> can you please investigate? Should
> MET be able to handle "string" variable attributes, like this?
>
> 	double lat(lat) ;
> 		string lat:long_name = "latitude" ;
> 		string lat:units = "degrees_north" ;
> 		string lat:standard_name = "latitude" ;
> 		lat:resolution = 0.04 ;
> 		lat:origin_offset = 0.02 ;
> 		string lat:bounds = "lat_bounds" ;
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/METplus/discussions/1416#discussioncomment-2128100>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AL3DNGTP77M7WXCZ3RO4D53U2AMYVANCNFSM5NS64BVA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
*Christopher J. Melick, PhD*

**Comment by hsoh-u** [comment, marked as answer]
[ CF - Convention of the standard_name ]:

https://cfconventions.org/cf-conventions/cf-conventions.html#standard-name

A standard name is associated with a variable via the attribute standard_name which takes **a string value** comprised of a standard name optionally followed by one or more blanks and a standard name modifier a string value from [Appendix C, Standard Name Modifiers](https://cfconventions.org/cf-conventions/cf-conventions.html#standard-name-modifiers).

[ Primitive data types of NetCDF file ]

https://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/build/ch02s02.html

NetCDF **does not support a character string type,** so these must be represented as character arrays. In this document, a one dimensional array of character data is simply referred to as a "string". An n-dimensional array of strings must be implemented as a character array of dimension (n,max_string_length), with the last (most rapidly varying) dimension declared large enough to contain the longest string in the array. All the strings in a given array are therefore defined to be equal in length. For example, an array of strings containing the names of the months would be dimensioned (12,9) in order to accommodate "September", the month with the longest name.


[ Support NC_STRING type for attributes at MET ]

1. The NC_STRING type is supported for the attributes which contains a string value.

https://github.com/dtcenter/MET/blob/develop/met/src/libcode/vx_nc_util/nc_utils.cc#L196
    Support both NC_CHAR and NC_STRING

2. If the numeric value is expected from the attribute. NC_STRING type is not supported yet.

https://github.com/dtcenter/MET/blob/develop/met/src/libcode/vx_nc_util/nc_utils.cc#L70
    Support only NC_CHAR

3. When attributes are copied from one NetCDF to another NetCDF, NC_STRING type attributes are not copied.

Note: 2 & 3 can be supported if needed.

**Reply by cjmelick** [reply]
![image](https://user-images.githubusercontent.com/49690266/153017388-e8c3da00-dae9-4a8b-a0d2-4a95ff5ce288.png)

John,

Your suggestion with the NOHRSC netCDF file worked but now I have a separate issue with the input model netCDF file which is complaining about that the variable needed 3 arguments but got 2.   I have a feeling it has something to do with the construct of the time.   Is the layout CF-compliant?   I have attached a screenshot for that section of the header.   Originally, wgrib2 generated the template and then I used the Iris library to update a post-processed parameter.  The time is not quite right -- it should be for the 24-hour period ending 3 Feb 2022.  Please let me know if you need more information.

Thanks,
Chris

**Reply by JohnHalleyGotway** [reply]
Chris,

This [Discussion Topic](https://github.com/dtcenter/METplus/discussions/954) describes how to send us data by posting it to our anonymous FTP site. So if you'd like me to take a closer look, feel free to send me a sample file.

But first, I'd recommend trying to run the plot_data_plane tool for this file:
```
plot_data_plane input_file.nc plot.ps 'name="var0_1_197_surface"; level="(0,*,*)"; set_attr_name="MyDataName";'
```

Where input_file.nc is the name of the NetCDF file whose header dump you posted. I want to point out 2 details here:

1. Note that my level setting "(0,*,*)" includes 1 entry for each dimension of the "var0_1_197_surface" NetCDF variable. The stars indicate the 2 dimensions which contain the gridded data (i.e. lat and lon) and the 0 requests the first time index value. Note that instead of "0" you could supply a time string in YYYYMMDD_HHMMSS format (e.g. level="(202008070_120000,*,*)";). That tells MET use the time index corresponding to that time stamp.

2. Note that I used "set_attr_name" to re-name the data being read. When reading this data into Point-Stat or Grid-Stat, the NetCDF variable name will be written to the FCST_VAR or OBS_VAR columns. I figure you may want to replace that NetCDF variable with a more descriptive name in the output. And that's how you'd do it.

Hope that helps.

**Reply by cjmelick** [reply]
John,

That very much helps.   The insight that paid off was I realized I needed the extra dimension to the Level for time.   Once I had your specification for level="(0,\*,\*)" in regrid_data_plane, I was able to get an output file!   I checked the spatial display of the snowfall accumulations in plot_data_plane before and after interpolation and the result worked just fine!   

My only follow up question would be for renaming the variable in regrid_data_plane -- how does that occur? 

Finally, is the string occurrence something that needs to be added to MET or is a change that really should occur upstream at NOAA when they are creating their data files for distribution?   

Thanks again,
Chris

**Reply by JohnHalleyGotway** [reply]
Chris,

You actually have 2 options in regrid_data_plane for renaming the variable in the output file:

1. Use the "-name" command line argument described in it's [usage statement](https://met.readthedocs.io/en/develop/Users_Guide/reformat_grid.html?highlight=regrid_data_plane#regrid-data-plane-usage).
Note that regrid_data_plane can process multiple fields and, if provided, the number of "-name" settings must match the number of variables being written.

For example, regrid 2 fields and explicitly specify the output variable names:
```
regrid_data_plane \
wrfprs_ruc13_12.tm00_G212 G001 regrid.nc \
-field 'name="TMP"; level="P850";' \
-field 'name="TMP"; level="Z2";'  \
-name T850,SFCTMP
```

The output file has these variables, as explicitly specified on the command line:
```
	float T850(lat, lon) ;
	float SFCTMP(lat, lon) ;
```

2. You could still use the "set_attr_name" and/or "set_attr_level" options in the config string, like this:
```
regrid_data_plane \
wrfprs_ruc13_12.tm00_G212 G001 regrid.nc \
-field 'name="TMP"; level="P850"; set_attr_name="T";' \
-field 'name="TMP"; level="Z2"; set_attr_name="T"; set_attr_level="SFC";' 
```

The output file has these variables, created as {name}_{level}:
```
	float T_P850(lat, lon) ;
	float T_SFC(lat, lon) ;
```

So either method works.

**Reply by JohnHalleyGotway** [reply]
As for the processing the NOHRSC NetCDF files, I'll write up a GitHub enhancement issue for the next release to handle those strings.

**Reply by cjmelick** [reply]
John,

Thanks for laying out the various options.   That's very useful!   I appreciate the assistance again.

Chris

**Reply by JohnHalleyGotway** [reply]
Chris, FYI, here's the new GitHub issue to enhance MET to handle this data:
https://github.com/dtcenter/MET/issues/2042

**Reply by cjmelick** [reply]
John, I took a look.  Now, if we can just get up to date with the right version of MET as well.   :-)

**Reply by JohnHalleyGotway** [reply]
Chris, I'll go ahead and mark this discussion as being answered and "Lock" conversation to prevent future posts. We found it difficult to keep track of new posts to previously answered discussions. So we "Lock" it to encourage users to post new questions to new discussions.

Thanks,
John
